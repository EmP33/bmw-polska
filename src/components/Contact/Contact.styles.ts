import styled from "styled-components";

export const Container = styled.div`
  position: relative;
  width: 100%;
  min-height: 100vh;
  margin: 3rem auto 0;
  display: grid;
  place-items: center;
  overflow: hidden;
`;

export const Contact = styled.nav`
  max-width: 750px;
  box-shadow: 3px 3px 60px #00000012;
  display: grid;
  background: var(--color-white);
  position: relative;
  margin: 5rem 0 10rem 0;

  @media only screen and (max-width: 600px) {
    box-shadow: none;
    margin: 20rem 0 25rem 0;
  }

  svg {
    position: absolute;
    z-index: 4;

    &:nth-of-type(1) {
      left: -350px;
      bottom: 175px;
    }
    &:nth-of-type(2) {
      left: -150px;
      top: 175px;
    }
    &:nth-of-type(3) {
      right: -300px;
      top: 200px;
      transform: rotate(250deg);
    }
    &:nth-of-type(4) {
      right: 50px;
      bottom: -125px;
      transform: rotate(-0deg);
    }
  }

  img {
    position: absolute;
    z-index: 5;

    &:nth-of-type(1) {
      right: -400px;
      top: 150px;

      @media only screen and (max-width: 600px) {
        left: -200px;
        top: -350px;
        width: 550px;
      }
    }
    &:nth-of-type(2) {
      left: -300px;
      bottom: 0;

      @media only screen and (max-width: 600px) {
        left: 30%;
        bottom: -350px;
        width: 550px;
      }
    }
  }

  & > .contact__header {
    position: relative;
    z-index: 10;
    background-color: var(--color-white);
    padding: 4rem 4rem 2rem 4rem;

    @media only screen and (max-width: 600px) {
      padding: 0rem 2rem 1rem 2rem;
    }

    h3 {
      font-family: var(--font-heading);
      font-size: 2.25rem;

      @media only screen and (max-width: 600px) {
        font-size: 1.5rem;
      }

      sup {
        font-size: 1rem;
      }
    }
    p {
      font-family: var(--font-primary);
      color: var(--color-gray-3);
      line-height: 2.5rem;
      font-size: 1.25rem;

      @media only screen and (max-width: 600px) {
        font-size: 1rem;
        line-height: 1.75rem;
      }
    }
  }

  & > .contact__form {
    position: relative;
    z-index: 10;
    background-color: var(--color-white);
    padding: 2rem 4rem 4rem 4rem;

    @media only screen and (max-width: 600px) {
      padding: 1rem 2rem 2rem 2rem;
    }
    h4 {
      font-size: 1.25rem;
      font-family: var(--font-heading);
      font-weight: 600;
    }

    form {
      margin-top: 1rem;
      display: grid;
      row-gap: 2rem;
      width: 75%;

      @media only screen and (max-width: 600px) {
        width: 100%;
        row-gap: 1rem;
      }

      [role="alert"] {
        color: red;
        font-family: var(--font-heading);
        font-size: 0.875rem;
      }

      .contact__radio-btns {
        display: grid;
        column-gap: 1.5rem;
        color: var(--color-gray-3);
        font-size: 1.25rem;
        grid-template-columns: repeat(2, max-content);

        /* The container */
        .container {
          position: relative;
          padding-left: 30px;
          margin-bottom: 12px;
          cursor: pointer;
          font-size: 22px;
          -webkit-user-select: none;
          -moz-user-select: none;
          -ms-user-select: none;
          user-select: none;

          @media only screen and (max-width: 600px) {
            font-size: 0.8125rem;
          }
        }

        /* Hide the browser's default checkbox */
        .container input {
          position: absolute;
          opacity: 0;
          cursor: pointer;
          height: 0;
          width: 0;
        }

        /* Create a custom checkbox */
        .checkmark {
          position: absolute;
          top: 50%;
          transform: translateY(-50%);
          left: 0;
          height: 22px;
          width: 22px;
          border-radius: 50%;
          background-color: var(--color-white);
          border: 2px solid var(--color-gray-6);
        }

        /* On mouse-over, add a grey background color */
        .container:hover input ~ .checkmark {
          background-color: var(--color-gray-4);
        }

        /* When the checkbox is checked, add a blue background */
        .container input:checked ~ .checkmark {
          background-color: var(--color-blue);
        }

        /* Create the checkmark/indicator (hidden when not checked) */
        .checkmark:after {
          content: "";
          position: absolute;
          display: none;
        }

        /* Show the checkmark when checked */
        .container input:checked ~ .checkmark:after {
          display: block;
        }

        /* Style the checkmark when checked */
        .container input:checked ~ .checkmark {
          border: none;
        }

        /* Style the checkmark/indicator */
        .container .checkmark:after {
          left: 7.5px;
          top: 5px;
          width: 4px;
          height: 8px;
          border: solid white;
          border-width: 0 3px 3px 0;
          -webkit-transform: rotate(45deg);
          -ms-transform: rotate(45deg);
          transform: rotate(45deg);
        }
      }

      input {
        width: 100%;
        padding: 1rem 0 1rem 1rem;
        color: var(--color-gray-1);
        font-size: 0.875rem;
        border: 1px solid var(--color-gray-4);

        @media only screen and (max-width: 600px) {
          font-size: 0.8125rem;
        }
      }

      button {
        justify-self: start;
      }
    }
  }
`;
