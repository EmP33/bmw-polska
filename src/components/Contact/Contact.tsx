import React from "react";
// Styles
import { PrimaryButton } from "../../Components.styles";
import { Container, Contact } from "./Contact.styles";
// Images
import frontBmw from "../../assets/images/bmw-ix-front.webp";
import sideBmw from "../../assets/images/13273-2022-bmw-ix.webp";
import { Path12, Path13, Path14, Path15 } from "../../assets";

// React hook form
import { useForm, SubmitHandler } from "react-hook-form";

type Inputs = {
  gender: string;
  name: string;
  surname: string;
  email: string;
  phone: string;
};
interface Props {
  handleGreetings: () => void;
}

const ContactComponent: React.FC<Props> = ({ handleGreetings }) => {
  const {
    register,
    handleSubmit,
    setValue,
    watch,
    formState: { errors },
  } = useForm<Inputs>();

  const onSubmit: SubmitHandler<Inputs> = (data) => {
    handleGreetings();

    setValue("gender", "");
    setValue("name", "");
    setValue("surname", "");
    setValue("email", "");
    setValue("phone", "");
  };

  return (
    <Container id="contact">
      <Contact>
        <img src={frontBmw} alt="bmw ix from front" data-aos="fade-right" />
        <Path12 />
        <Path13 />
        <Path14 />
        <Path15 />
        <div className="contact__header">
          <h3>
            UMÓW JAZDĘ PRÓBNĄ <sup>BMW iX</sup>
          </h3>
          <p>
            Zachęcamy do ograniczenia poruszania się i pozostania w domu. <br />
            Jeśli jednak jazda testowa okaże się niezbędna prosimy o wypełnienie
            poniższego formularza.
          </p>
        </div>
        <div className="contact__form">
          <h4>Wypełnij formularz swoimi danymi.</h4>
          <form onSubmit={handleSubmit(onSubmit)}>
            <div className="contact__radio-btns">
              <label htmlFor="mr" className="container">
                <input
                  type="radio"
                  id="mr"
                  value="he"
                  {...register("gender", { required: true })}
                />
                <span
                  tabIndex={0}
                  aria-checked={watch("gender") === "he"}
                  role="radio"
                  className="checkmark"
                ></span>
                Pan
              </label>
              <label htmlFor="mrs" className="container">
                <input
                  type="radio"
                  id="mrs"
                  value="she"
                  {...register("gender", { required: true })}
                />
                <span
                  tabIndex={0}
                  aria-checked={watch("gender") === "she"}
                  role="radio"
                  className="checkmark"
                ></span>
                Pani
              </label>
              {errors.gender && <span role="alert">Wybierz płeć</span>}
            </div>
            <div className="input-wrapper">
              <input
                type="text"
                placeholder="Imię*"
                {...register("name", { required: true })}
              />
              {errors.name && (
                <span role="alert">Pole imię nie może być puste.</span>
              )}
            </div>
            <div className="input-wrapper">
              <input
                type="text"
                placeholder="Nazwisko*"
                {...register("surname", { required: true })}
              />
              {errors.surname && (
                <span role="alert">Pole nazwisko nie może być puste.</span>
              )}
            </div>
            <div className="input-wrapper">
              <input
                type="text"
                placeholder="E-mail*"
                {...register("email", {
                  required: true,
                  pattern: /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/i,
                })}
              />
              {errors.email && (
                <span role="alert">Wprowadź poprawny adres email.</span>
              )}
            </div>
            <div className="input-wrapper">
              <input
                type="number"
                placeholder="Numer telefonu*"
                {...register("phone", { required: true, minLength: 9 })}
              />
              {errors.phone && (
                <span role="alert">Wprowadź poprawny numer.</span>
              )}
            </div>
            <PrimaryButton type="submit">WYŚLIJ</PrimaryButton>
          </form>
        </div>
        <img src={sideBmw} alt="bmw ix from side" data-aos="fade-left" />
      </Contact>
    </Container>
  );
};

export default ContactComponent;
