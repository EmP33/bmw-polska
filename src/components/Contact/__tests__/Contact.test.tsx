import Contact from "../Contact";
import ReactShallowRenderer from "react-test-renderer/shallow";
import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("<Contact />", () => {
  it("should render properly", () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<Contact handleGreetings={jest.fn()} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("should show no errors by default", () => {
    render(<Contact handleGreetings={jest.fn()} />);
    const errorMessages = screen.queryAllByTestId("error");
    expect(errorMessages).toStrictEqual([]);
  });

  it("should show errors when submit button is clicked without any values", async () => {
    render(<Contact handleGreetings={jest.fn()} />);
    const submitButton = screen.getByRole("button", { name: "WYŚLIJ" });

    userEvent.click(submitButton);
    expect(await screen.findAllByRole("alert")).toHaveLength(5);
  });

  it("should display email error when email is invalid", async () => {
    render(<Contact handleGreetings={jest.fn()} />);
    const submitButton = screen.getByRole("button", { name: "WYŚLIJ" });
    const emailInput = screen.getByPlaceholderText("E-mail*");

    userEvent.type(emailInput, "examplegmail.com");
    userEvent.click(submitButton);

    expect(
      await screen.findByText("Wprowadź poprawny adres email.")
    ).toBeInTheDocument();
  });

  it("should display phone error when phone is invalid", async () => {
    render(<Contact handleGreetings={jest.fn()} />);
    const submitButton = screen.getByRole("button", { name: "WYŚLIJ" });
    const phoneInput = screen.getByPlaceholderText("Numer telefonu*");

    userEvent.type(phoneInput, "12345");
    userEvent.click(submitButton);

    expect(
      await screen.findByText("Wprowadź poprawny numer.")
    ).toBeInTheDocument();
  });

  it("should be no error when all input values are correct", async () => {
    render(<Contact handleGreetings={jest.fn()} />);
    const submitButton = screen.getByRole("button", { name: "WYŚLIJ" });
    const radioButton = screen.getByLabelText("Pan");
    const nameInput = screen.getByPlaceholderText("Imię*");
    const surnameInput = screen.getByPlaceholderText("Nazwisko*");
    const emailInput = screen.getByPlaceholderText("E-mail*");
    const phoneInput = screen.getByPlaceholderText("Numer telefonu*");
    userEvent.click(radioButton);
    userEvent.type(nameInput, "John");
    userEvent.type(surnameInput, "Doe");
    userEvent.type(phoneInput, "666555999");
    userEvent.type(emailInput, "johndoe@gmail.com");
    userEvent.click(submitButton);

    expect(screen.queryByRole("alert")).toBe(null);
  });
});
