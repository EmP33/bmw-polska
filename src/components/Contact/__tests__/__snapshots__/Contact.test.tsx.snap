// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact /> should render properly 1`] = `
<styled.div
  id="contact"
>
  <styled.nav>
    <img
      alt="bmw ix from front"
      src="bmw-ix-front.png"
    />
    <Path12 />
    <Path13 />
    <Path14 />
    <Path15 />
    <div
      className="contact__header"
    >
      <h3>
        UMÓW JAZDĘ PRÓBNĄ 
        <sup>
          BMW iX
        </sup>
      </h3>
      <p>
        Zachęcamy do ograniczenia poruszania się i pozostania w domu. 
        <br />
        Jeśli jednak jazda testowa okaże się niezbędna prosimy o wypełnienie poniższego formularza.
      </p>
    </div>
    <div
      className="contact__form"
    >
      <h6>
        Wypełnij formularz swoimi danymi.
      </h6>
      <form
        onSubmit={[Function]}
      >
        <div
          className="contact__radio-btns"
        >
          <label
            className="container"
            htmlFor="mr"
          >
            <input
              id="mr"
              name="gender"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="he"
            />
            <span
              className="checkmark"
            />
            Pan
          </label>
          <label
            className="container"
            htmlFor="mrs"
          >
            <input
              id="mrs"
              name="gender"
              onBlur={[Function]}
              onChange={[Function]}
              type="radio"
              value="she"
            />
            <span
              className="checkmark"
            />
            Pani
          </label>
        </div>
        <div
          className="input-wrapper"
        >
          <input
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Imię*"
            type="text"
          />
        </div>
        <div
          className="input-wrapper"
        >
          <input
            name="surname"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Nazwisko*"
            type="text"
          />
        </div>
        <div
          className="input-wrapper"
        >
          <input
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="E-mail*"
            type="text"
          />
        </div>
        <div
          className="input-wrapper"
        >
          <input
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Numer telefonu*"
            type="number"
          />
        </div>
        <styled.button
          type="submit"
        >
          WYŚLIJ
        </styled.button>
      </form>
    </div>
    <img
      alt="bmw ix from side"
      src="13273-2022-bmw-ix.png"
    />
  </styled.nav>
</styled.div>
`;
